lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -0,0 +1,374 @@
+                UsdSceneItem::create(MayaUsd::ufe::stagePath(stage), stage->GetPseudoRoot()),
+                kDefaultMaterialScopeName,
+                "Scope");
+            _createScopeCmd->execute();
+            scopePath = _createScopeCmd->newUfePath();
+            // The code automatically appends a "1". We need to rename:
+            auto itemOps = Ufe::SceneItemOps::sceneItemOps(Ufe::Hierarchy::createItem(scopePath));
+            auto rename = itemOps->renameItemCmd(Ufe::PathComponent(kDefaultMaterialScopeName));
+            _renameScopeCmd = rename.undoableCommand;
+            scopePath = rename.item->path();
+        }
+    }
+    if (scopePath.empty()) {
+        TF_RUNTIME_ERROR("Failed to create materials scope at stage root");
+        return;
+    }
+    //
+    // 2. Create the Material if it does not exist:
+    //
+    PXR_NS::SdrRegistry&          registry = PXR_NS::SdrRegistry::GetInstance();
+    PXR_NS::SdrShaderNodeConstPtr shaderNodeDef
+        = registry.GetShaderNodeByIdentifier(TfToken(_nodeId));
+    if (!shaderNodeDef) {
+        TF_RUNTIME_ERROR("Unknown shader identifier: %s", _nodeId.c_str());
+        return;
+    }
+    if (shaderNodeDef->GetOutputNames().empty()) {
+        TF_RUNTIME_ERROR("Surface shader %s does not have any outputs", _nodeId.c_str());
+        return;
+    }
+    auto scopeItem = std::dynamic_pointer_cast<UsdSceneItem>(Ufe::Hierarchy::createItem(scopePath));
+    _createMaterialCmd = UsdUndoAddNewPrimCommand::create(
+        scopeItem, shaderNodeDef->GetFamily().GetString(), "Material");
+    _createMaterialCmd->execute();
+    if (!_createMaterialCmd->newPrim()) {
+        TF_RUNTIME_ERROR("Could not create material under %s", scopePath.string().c_str());
+        return;
+    }
+    //
+    // 3. Create the Shader if it does not exist:
+    //
+    auto materialItem = std::dynamic_pointer_cast<UsdSceneItem>(
+        Ufe::Hierarchy::createItem(_createMaterialCmd->newUfePath()));
+    _createShaderCmd = UsdUndoCreateFromNodeDefCommand::create(
+        shaderNodeDef, materialItem, shaderNodeDef->GetFamily().GetString());
+    _createShaderCmd->execute();
+    if (!_createShaderCmd->insertedChild()) {
+        TF_RUNTIME_ERROR(
+            "Could not create %s shader under %s",
+            _nodeId.c_str(),
+            _createMaterialCmd->newUfePath().string().c_str());
+        return;
+    }
+    //
+    // 4. Connect the Shader to the material:
+    //
+    if (!connectShaderToMaterial(
+            _createShaderCmd->insertedChild(), _createMaterialCmd->newPrim())) {
+        return;
+    }
+    //
+    // 5. Bind the material to the parent primitive:
+    //
+    _bindCmd = std::make_shared<BindMaterialUndoableCommand>(
+        parentItem->prim(), materialItem->prim().GetPath());
+    _bindCmd->execute();
+
+    // Successfully completed the task:
+    guard.markAsCompleted();
+}
+
+void UsdUndoAssignNewMaterialCommand::undo()
+{
```

Comment:
That is a very good question, and my answer is: try to complete the undo even if an intermediate step fails.

This is what this code is doing. Since "undo" returns "void", we do not have visibility on the success or failure of the individual calls.

Reply:
Command undo failure should be reported through exceptions --- and they should be truly exceptional, because if undo fails the undo stack should be considered corrupt, and cleared, because redo won't bring state back to where it was.  I would expect undo failure to be a logic error, rather than a run-time error, in C++ exception terms.